### ISSUE TITLE
(Baturay) As a developer, I would like to implement a component to handle scheduled shutdowns

### ISSUE THREAD

**Issue by cagdasgerede** [tobbetu-ceng-industry-collaborations/smarteff-main#5]
We would like to give the users some time to take action (save their work, or request suspension of the shutdown) before their devices are turned off. Therefore, the actual shutdown should be scheduled for, for example, 5 minutes later.

If a suspension request exists for a device already, then no shutdown should be scheduled. (Talk to developer of #4 to agree on the details of how to represent a suspension in the database)

When a shutdown is scheduled, the device users should be notified with a notification so that the users have a small window of time to take action before the shutdown. (Talk to developer of #9 and make the necessary accomodations to make both components compatible with each other)

A background process should periodically check for scheduled shutdowns and notice that a scheduled shutdown is ready to be processed. At this point, the component should check if there is a suspension for the device. If there is, then the shutdown should be cancelled.  Otherwise, an event should be fired to turn off the device. The actual messaging needed to shut down the device will be handled by another component. So this component should only fire an appropriate event to request the device to shutdown and the other component will take care of the actual device communication.

# Simulation mode
Examine #11 and make the necessary accomodations to run the component in a simulation friendly way.

**Comment by BaturayOnural** [comment]
I would like to explain workflow of our scheduler system described in #5

## Schedule
We have a scheduler called **schedule.py** which can be runned as:

![Scheduled](https://user-images.githubusercontent.com/20816596/75889581-32b61b00-5e3e-11ea-9dff-9783d9184975.png)
The commands above creates 4 background processes which will:

i.  Terminate Device 2 in 1500 seconds
ii. Terminate Device 3 in 1500 seconds
iii.Terminate Device 4 in 1700 seconds
iv.Terminate Device 5 in 2100 seconds

the syntax is as follows: (& at the end indicates that it is a background process)
```sh
$ python3 schedule.py <scheduledShutdownID> <deviceID> <secondsUntilShutdown> &
```
We can validate that these processes are actually queued for shutdowns via terminal command ps ax | grep schedule.py

![scheduledShutdowns](https://user-images.githubusercontent.com/20816596/75890780-e370ea00-5e3f-11ea-9be8-e16f8d4c5c45.png)

## de-Schedule

We have a de-scheduler called **de_schedule.py** which can be runned as:

![Kill-1](https://user-images.githubusercontent.com/20816596/75891261-94778480-5e40-11ea-800b-1253ffbde4a0.png)

The command above kills the process that has been called with given **scheduledShutdown ID**

the syntax is as follows: 
```sh
$ python3 de_schedule.py <scheduledShutdownID>
```
We can validate the process we intended to terminate is actually terminated via terminal command ps ax | grep schedule.py

![result](https://user-images.githubusercontent.com/20816596/75891901-7ceccb80-5e41-11ea-84de-03f03d5bbeed.png)

Obviously, **we do not see** the automated shutdown process with **scheduledShutdownID=1**

We will be using these scripts in our main app components which are mainly implemented in #4, in cases of scheduling-descheduling needs.

**Comment by BaturayOnural** [comment]
I am closing the issue since it has completely implemented as described. Feel free to write about any unexpected behavior of suspensions and scheduled shutdowns under this thread.
